### PR TITLE
adding the github_ca parameter to ensure that when we have a GitHub i…

### PR DESCRIPTION
**Pull request by ckyriakidou**
…dentity provider, it can function as well as the openid and ldap ones
https://github.com/openshift/openshift-ansible/issues/10565

Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1647793

### CODE REVIEW

File: roles/openshift_master_facts/tasks/main.yml [hunk 1]
```diff
@@ -43,6 +43,7 @@
       session_name: "{{ openshift_master_session_name | default(None) }}"
       ldap_ca: "{{ openshift_master_ldap_ca | default(lookup('file', openshift_master_ldap_ca_file) if openshift_master_ldap_ca_file is defined else None) }}"
       openid_ca: "{{ openshift_master_openid_ca | default(lookup('file', openshift_master_openid_ca_file) if openshift_master_openid_ca_file is defined else None) }}"
+      github_ca: "{{ openshift_master_github_ca | default(lookup('file', openshift_master_github_ca_file) if openshift_master_github_ca_file is defined else None) }}"
```
Comment (vrutkovs) [inline roles/openshift_master_facts/tasks/main.yml]: This var needs to be mentioned in `hosts.example`

File: roles/openshift_control_plane/tasks/main.yml [hunk 2]
```diff
@@ -82,6 +82,17 @@
...
 
+- name: Create the GitHub (Enterprise) ca file if needed
+  copy:
+    dest: "/etc/origin/master/{{ item.name }}_github_ca.crt"
+    content: "{{ openshift.master.github_ca }}"
+    mode: 0600
+    backup: yes
+  when:
+  - openshift.master.github_ca is defined
+  - item.kind == 'GitHubIdentityProvider'
```
Comment (vrutkovs) [inline roles/openshift_control_plane/tasks/main.yml]: `roles/lib_utils/filter_plugins/openshift_master.py` needs to be updated to have this class

Comment (mjudeikis) [reply roles/openshift_control_plane/tasks/main.yml]: Any ref to this?

Comment (vrutkovs) [reply roles/openshift_control_plane/tasks/main.yml]: Oh, nvm, its already there

Comment (vrutkovs) [reply roles/openshift_control_plane/tasks/main.yml]: Not entirely sure how Github Enterprise works though - I guess it needs a custom URL too?

Comment (ckyriakidou) [reply roles/openshift_control_plane/tasks/main.yml]: hostname according to the https://docs.okd.io/3.11/install_config/configuring_authentication.html#GitHub documentation

File: roles/openshift_master_facts/tasks/main.yml (hunk 1 above)
Comment (ckyriakidou) [reply roles/openshift_master_facts/tasks/main.yml]: @vrutkovs done, with all the missing GitHubIdentityProvider missing examples too.

File: roles/lib_utils/filter_plugins/openshift_master.py [hunk 3]
```diff
@@ -454,13 +454,17 @@ class GitHubIdentityProvider(IdentityProviderOauthBase):
...
+                           ['teams'],
+                           ['ca'],
+                           ['hostname']]
 
     def validate(self):
         ''' validate this idp instance '''
         if self.challenge:
             raise errors.AnsibleFilterError("|failed provider {0} does not "
                                             "allow challenge authentication".format(self.__class__.__name__))
+        if self._idp['ca'] != "":
```
Comment (vrutkovs) [inline roles/lib_utils/filter_plugins/openshift_master.py]: I don't think it needs to be conditional. See other providers implementation, https://github.com/openshift/openshift-ansible/issues/9397#issuecomment-413158710 and https://bugzilla.redhat.com/show_bug.cgi?id=1633137

Comment (ckyriakidou) [reply roles/lib_utils/filter_plugins/openshift_master.py]: I can remove it, do you think there should be a conditional checking if the file is there/exists? If not the installation will fail (when masters get restarted).

Comment (vrutkovs) [reply roles/lib_utils/filter_plugins/openshift_master.py]: If the admin doesn't use `_ca` or `_ca_file` vars then its their responsibility to have this file present there - a different playbook, host image etc.

**Comment by ckyriakidou** [comment]
@vrutkovs all done.

**Comment by vrutkovs** [comment]
@ckyriakidou seems 3.11 and 3.10 cherrypicks should be done manually. Do you need help with that?

**Comment by ckyriakidou** [comment]
@vrutkovs I think so, do you have any docs on it? i.e. I know the ansible repo does /backport/branch/issue iirc and then cherrypick. What's the process here?

**Comment by vrutkovs** [comment]
Yeah, this is what we do as well, however automatic cherrypick has failed due to merge conflicts.

This needs to be done manually. Please make a branch of `release-3.10`, cherrypick this commit, resolve merge conflict and make a new pull-request to `release-3.10`. In the PR description please mention this is a cherrypick of https://github.com/openshift/openshift-ansible/pull/10566

See https://wiki.koha-community.org/wiki/Using_Git_Cherry_Pick for a guide about cherrypicking and merge confict resolution.

**Pull request merged**